torch/csrc/jit/codegen/cuda/predicate_compute.cpp

```diff
@@ -52,8 +52,108 @@ bool isOutputLocal(const kir::Expr* expr) {
       });
 }
 
+constexpr std::array<ParallelType, 6> parallel_types = {
+    ParallelType::BIDx,
+    ParallelType::BIDy,
+    ParallelType::BIDz,
+    ParallelType::TIDx,
+    ParallelType::TIDy,
+    ParallelType::TIDz};
+
 } // namespace
 
+UnswitchPredicateKey::UnswitchPredicateKey() : id_(nullptr) {
+  for (auto pt : parallel_types) {
+    parallel_ids_.insert({pt, nullptr});
+  }
+}
+
+// For a predicated concrete domain, id, find which thread parallel
+// types are used. For each used parallel type, find the concrete
+// domain that the paralllel type is associated with. The parallelized
+// concrete domains are used to uniquely collect all necessary
+// unswitch predicates.
+UnswitchPredicateKey::UnswitchPredicateKey(
+    IterDomain* id,
+    const ReferenceTensor& reference)
+    : id_(id) {
+  // Initialize the parallelized domain map
+  for (auto pt : parallel_types) {
+    parallel_ids_.insert({pt, nullptr});
+  }
+
+  // The id parameter is a concrete domain. Needs to find the
+  // corresponding reference domain to find leaf domains that are
+  // parallelized.
+  IterDomain* ref_id = reference.concrete_to_id.at(id_);
+  TensorDomain* ref_td = reference.domain;
+
+  std::vector<Val*> parallelized_leaf_domains;
+  std::copy_if(
+      ref_td->domain().begin(),
+      ref_td->domain().end(),
+      std::back_inserter(parallelized_leaf_domains),
+      [](IterDomain* x) { return isParallelTypeThread(x->getParallelType()); });
+
+  // If the reference is not parallelized at all, no need to
+  // differentiate keys based on how id_ is parallelized
+  if (parallelized_leaf_domains.empty()) {
+    return;
+  }
+
+  auto all_vals =
+      DependencyCheck::getAllValsBetween({ref_id}, parallelized_leaf_domains);
+  std::vector<IterDomain*> leaf_ids;
+  std::copy_if(
+      ref_td->domain().begin(),
+      ref_td->domain().end(),
+      std::back_inserter(leaf_ids),
+      [&](IterDomain* x) {
+        return std::find(all_vals.begin(), all_vals.end(), x) != all_vals.end();
+      });
+
+  if (leaf_ids.empty()) {
+    // None of the parallelized ids are derived from ref_id
+    return;
+  }
+
+  // Find the corresponding concrete id for each parallel type
+  for (auto ref_leaf : leaf_ids) {
+    auto pt = ref_leaf->getParallelType();
+    auto it = std::find_if(
+        reference.concrete_to_id.begin(),
+        reference.concrete_to_id.end(),
+        [&](const auto& kv) { return kv.second == ref_leaf; });
+    TORCH_INTERNAL_ASSERT(it != reference.concrete_to_id.end());
+    auto concrete_leaf = it->first;
+    parallel_ids_.at(pt) = concrete_leaf;
+  }
```

Comment:
Wouldn't this have been easier to do by just flipping the `concrete_to_id` map? Seems you could have then just looked for the parallelized leaf domains in the reference, directly grabbed the concrete leaf, and thrown it in the parlalel_ids_ map. You wouldn't even have to flip the whole map, just those in the reference domain.

Conversely I think it would be reasonable to just put `id_to_concrete` map into the reference domain, even if this is the only place we use it.

Reply:
I'll add `id_to_concrete`. Thanks for the suggesion.